### PR TITLE
[DeviceSanitizer] Support detecting out-of-bounds errors on sycl::buffer

### PR DESCRIPTION
**Pull request by zhaomaosu**

### CODE REVIEW

File: source/loader/layers/sanitizer/asan_interceptor.hpp [hunk 1]
```diff
@@ -79,6 +80,12 @@ struct QueueInfo {
     }
 };
 
+struct KernelInfo {
+    ur_kernel_handle_t Handle;
+    ur_shared_mutex Mutex;
+    std::unordered_map<int, std::shared_ptr<MemBuffer>> BufferArgs;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.hpp:86]:
```suggestion
    std::unordered_map<uint32_t, std::shared_ptr<MemBuffer>> BufferArgs;
```

File: source/loader/layers/sanitizer/asan_interceptor.hpp [hunk 2]
```diff
@@ -79,6 +80,12 @@ struct QueueInfo {
     }
 };
 
+struct KernelInfo {
+    ur_kernel_handle_t Handle;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.hpp:84]: Add constructor & deconstructor, "urKernelRetain" and "urKernelRelease".

File: source/loader/layers/sanitizer/asan_interceptor.hpp [hunk 3]
```diff
@@ -173,6 +180,20 @@ class SanitizerInterceptor {
...
+
+    ur_result_t eraseMemBuffer(ur_mem_handle_t MemHandle);
+
+    std::shared_ptr<MemBuffer> getMemBuffer(ur_mem_handle_t MemHandle);
+
+    std::shared_ptr<KernelInfo> getKernelInfo(ur_kernel_handle_t Kernel) {
+        std::shared_lock<ur_shared_mutex> Guard(m_KernelMapMutex);
+        assert(m_KernelMap.find(Kernel) != m_KernelMap.end());
+        return m_KernelMap[Kernel];
+    }
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.hpp:195]: nit: move "getKernelInfo" below "getContextInfo"

File: source/loader/layers/sanitizer/asan_interceptor.hpp [hunk 4]
```diff
@@ -173,6 +180,20 @@ class SanitizerInterceptor {
                              std::shared_ptr<DeviceInfo> &CI);
     ur_result_t eraseDevice(ur_device_handle_t Device);
 
+    ur_result_t insertKernel(ur_kernel_handle_t Kernel);
+
+    ur_result_t insertMemBuffer(std::shared_ptr<MemBuffer> MemBuffer);
+
+    ur_result_t eraseMemBuffer(ur_mem_handle_t MemHandle);
+
+    std::shared_ptr<MemBuffer> getMemBuffer(ur_mem_handle_t MemHandle);
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.hpp:189]: nit: remove blank line between them.

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 5]
```diff
@@ -316,6 +316,802 @@ __urdlllocal ur_result_t UR_APICALL urContextRelease(
...
+
+    if (nullptr == pfnBufferCreate) {
+        return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    }
+
+    context.logger.debug("==== urMemBufferCreate");
+
+    char *hostPtrOrNull = (flags & UR_MEM_FLAG_USE_HOST_POINTER)
+                              ? ur_cast<char *>(pProperties->pHost)
+                              : nullptr;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:340]: I'm not sure, should you check "pProperties" is nullptr?

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 6]
```diff
@@ -316,6 +316,802 @@ __urdlllocal ur_result_t UR_APICALL urContextRelease(
...
+
+    std::shared_ptr<MemBuffer> pMemBuffer =
+        std::make_shared<MemBuffer>(hContext, size, hostPtrOrNull);
+    ur_result_t result = context.interceptor->insertMemBuffer(pMemBuffer);
+    *phBuffer = ur_cast<ur_mem_handle_t>(pMemBuffer.get());
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:345]: should you check if “phBuffer” is nullptr?

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 7]
```diff
@@ -316,6 +316,802 @@ __urdlllocal ur_result_t UR_APICALL urContextRelease(
...
+        UrReturnHelper ReturnValue(propSize, pPropValue, pPropSizeRet);
+        switch (propName) {
+        case UR_MEM_INFO_CONTEXT: {
+            return ReturnValue(MemBuffer->Context);
+        }
+        case UR_MEM_INFO_SIZE: {
+            return ReturnValue(size_t{MemBuffer->Size});
+        }
+        default: {
+            die("urMemGetInfo: Parameter is not implemented");
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:385]: IMO, this situation should "return" not "die".

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 8]
```diff
@@ -316,6 +316,802 @@ __urdlllocal ur_result_t UR_APICALL urContextRelease(
...
+
+    if (nullptr == pfnBufferPartition) {
+        return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    }
+
+    context.logger.debug("==== urMemBufferPartition");
+
+    if (auto ParentBuffer = context.interceptor->getMemBuffer(hBuffer)) {
+        std::shared_ptr<MemBuffer> SubBuffer = std::make_shared<MemBuffer>(
+            ParentBuffer, pRegion->origin, pRegion->size);
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:466]: In some cases creating subbuffer can fail, this due to invalid pRegion(larger than size of parent buffer), is it your responsibility to check as well?

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 9]
```diff
@@ -316,6 +316,802 @@ __urdlllocal ur_result_t UR_APICALL urContextRelease(
...
+            USMDesc.align = MemBuffer->getAlignment();
+            ur_usm_pool_handle_t Pool{};
+            UR_CALL(context.interceptor->allocateMemory(
+                Context, nullptr, &USMDesc, Pool, size, AllocType::HOST_USM,
+                ppRetMap));
+        }
+
+        // Actually, if the access mode is write only, we don't need to do this
+        // copy. However, in that way, we cannot generate a event to user. So,
+        // we'll aways do copy here.
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:950]: maybe you can use "urEnqueueEventsWait" to create a user event?

File: source/loader/layers/sanitizer/ur_sanddi.cpp [hunk 10]
```diff
@@ -434,6 +1296,19 @@ __urdlllocal ur_result_t UR_APICALL urGetEnqueueProcAddrTable(
 
     ur_result_t result = UR_RESULT_SUCCESS;
 
+    pDdiTable->pfnKernelLaunch = ur_sanitizer_layer::urEnqueueKernelLaunch;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/ur_sanddi.cpp:1299]: Duplication

File: source/loader/layers/sanitizer/asan_interceptor.hpp (hunk 1 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.hpp]: Done

File: source/loader/layers/sanitizer/asan_interceptor.hpp [hunk 11]
```diff
@@ -173,6 +180,20 @@ class SanitizerInterceptor {
                              std::shared_ptr<DeviceInfo> &CI);
     ur_result_t eraseDevice(ur_device_handle_t Device);
 
+    ur_result_t insertKernel(ur_kernel_handle_t Kernel);
```
[the remark this answers is not in the document]
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.hpp:195]: Done

File: source/loader/layers/sanitizer/asan_interceptor.hpp (hunk 2 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.hpp:84]: Done

File: source/loader/layers/sanitizer/asan_interceptor.hpp (hunk 3 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.hpp:209]: Done

File: source/loader/layers/sanitizer/asan_interceptor.hpp (hunk 4 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.hpp]: Done

File: source/loader/layers/sanitizer/asan_interceptor.hpp [hunk 12]
```diff
@@ -210,10 +231,15 @@ class SanitizerInterceptor {
...
-
     std::unordered_map<ur_device_handle_t, std::shared_ptr<DeviceInfo>>
         m_DeviceMap;
     ur_shared_mutex m_DeviceMapMutex;
+    std::unordered_map<ur_kernel_handle_t, std::shared_ptr<KernelInfo>>
+        m_KernelMap;
+    ur_shared_mutex m_KernelMapMutex;
+    std::unordered_map<ur_mem_handle_t, std::shared_ptr<MemBuffer>>
+        m_MemBufferMap;
+    ur_shared_mutex m_MemBufferMapMutex;
```
[the remark this answers is not in the document]
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.hpp]: Done

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 5 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp]: Done

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 6 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:350]: According to UR spec, the phBuffer is not an optional argument, so I think there is no need to check null for it

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 7 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp]: Done

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 6 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:354]: Added the check...

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 8 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:478]: Done

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 10 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/ur_sanddi.cpp]: Done

File: source/loader/layers/sanitizer/ur_sanddi.cpp (hunk 9 above)
Comment (AllanZyne) [reply source/loader/layers/sanitizer/ur_sanddi.cpp:962]: not important

**Pull request closed without merging**